src/v/cloud_storage/partition_manifest.cc

```diff
@@ -2001,67 +2001,18 @@ partition_manifest::timequery(model::timestamp t) const {
 
-    // From this point on, we have finite positive differences between
-    // base and max for both offset and time.
-    model::offset interpolated_offset = model::offset{std::lerp(
-      base_offset(),
-      max_offset(),
-      float(t() - base_t()) / float(max_t() - base_t()))};
-
-    vlog(
-      cst_log.debug, "timequery t={} offset guess {}", t, interpolated_offset);
-
-    // 2. Convert offset guess into segment guess.  This is not a strictly
-    // correct offset lookup, we just want something close.
-    auto segment_iter = [&] {
-        auto candidate = _segments.lower_bound(interpolated_offset);
-        if (candidate != _segments.end()) {
-            return candidate;
-        }
-        return _segments.at_index(_segments.size() - 1);
-    }();
-    vlog(
-      cst_log.debug,
-      "timequery t={} segment initial guess {}",
-      t,
-      *segment_iter);
-
-    // 3. Do linear search backward or forward from the location
-    //    we probed, to find the right segment.
-    if (segment_iter->base_timestamp < t) {
-        // Our guess's base_timestamp is before the search point, so our
-        // result must be this segment or a later one: search forward
-        const auto& end = _segments.at_index(_segments.size() - 1);
-        for (; segment_iter != end; ++segment_iter) {
-            const auto& s = *segment_iter;
-            // We found a segment bounding the search point - or -
-            // We found the first segment after the search point
-            if (s.max_timestamp >= t || s.base_timestamp > t) {
-                return s;
-            }
-        }
-        vassert(
-          end != _segments.end(), "Trying to dereference invalid iterator");
-        return *end;
-    } else {
-        // Search backwards: we must peek at each preceding segment
-        // to see if its max_timestamp is >= the query t, before
-        // deciding whether to walk back to it.
-        vlog(cst_log.debug, "timequery t={} reverse {}", t, *segment_iter);
-
-        while (segment_iter != _segments.begin()) {
-            vlog(cst_log.debug, "timequery t={} reverse {}", t, *segment_iter);
-            // NOTE segment_iter is a ForwardIterator. backward walking is
-            // implemented in this slightly awkward way
-            auto prev = _segments.prev(segment_iter);
-            if (prev->max_timestamp >= t) {
-                segment_iter = std::move(prev);
-            } else {
-                break;
-            }
+    auto base_timestamp = _segments.get_base_timestamp_column().begin();
+    auto max_timestamp = _segments.get_max_timestamp_column().begin();
+    size_t target_ix = 0;
+    while (!base_timestamp.is_end()) {
+        if (*max_timestamp >= t.value() || *base_timestamp > t.value()) {
+            target_ix = base_timestamp.index();
+            break;
         }
-        return *segment_iter;
+        ++base_timestamp;
+        ++max_timestamp;
     }
+    return *_segments.at_index(target_ix);
 }
```

Comment:
Could you update the comment on this function please? So this turned out being faster? It's not entirely obvious to me since the previous version did less steps, but had a worse memory access pattern.

Reply:
_segments.prev increments the begin iterator until it reaches the correct value, so on paper is O(n²), util iteration segment_meta_cstore is fixed